File: api/state.py
from collections import deque

from lib.feature_engineering import RequestRecord

WINDOW_MAX = 300.0

_ip_records: dict[str, deque[RequestRecord]] = {}


def add_record(ip: str, record: RequestRecord) -> None:
    if ip not in _ip_records:
        _ip_records[ip] = deque()
    _ip_records[ip].append(record)
    _cleanup(ip, record.timestamp)


def get_records(ip: str, now: float, window: float) -> list[RequestRecord]:
    if ip not in _ip_records:
        return []
    cutoff = now - window
    return [r for r in _ip_records[ip] if r.timestamp >= cutoff]


def _cleanup(ip: str, now: float) -> None:
    q = _ip_records[ip]
    cutoff = now - WINDOW_MAX
    while q and q[0].timestamp < cutoff:
        q.popleft()
    if not q:
        del _ip_records[ip]

File: ultra-low-latency/server/src/state.rs
use std::collections::{HashMap, VecDeque};
use std::sync::Mutex;

const WINDOW_MAX: f64 = 300.0;

#[derive(Clone)]
pub struct RequestRecord {
    pub timestamp: f64,
    pub endpoint: String,
    pub status_code: i32,
    pub payload_size: f64,
    pub user_agent: String,
    pub response_time: f64,
}

pub struct IpState {
    records: Mutex<HashMap<String, VecDeque<RequestRecord>>>,
}

impl IpState {
    pub fn new() -> Self {
        Self {
            records: Mutex::new(HashMap::new()),
        }
    }

    pub fn add_record(&self, ip: &str, record: RequestRecord) {
        let mut map = self.records.lock().unwrap();
        let q = map.entry(ip.to_string()).or_insert_with(VecDeque::new);
        let now = record.timestamp;
        q.push_back(record);

        let cutoff = now - WINDOW_MAX;
        while let Some(front) = q.front() {
            if front.timestamp < cutoff {
                q.pop_front();
            } else {
                break;
            }
        }
        if q.is_empty() {
            map.remove(ip);
        }
    }

    pub fn get_records(&self, ip: &str, now: f64, window: f64) -> Vec<RequestRecord> {
        let map = self.records.lock().unwrap();
        match map.get(ip) {
            None => Vec::new(),
            Some(q) => {
                let cutoff = now - window;
                q.iter()
                    .filter(|r| r.timestamp >= cutoff)
                    .cloned()
                    .collect()
            }
        }
    }
}
